### ISSUE TITLE
abort while optimizing emscripten-compiled code

### ISSUE THREAD

**Issue by kumpera** [WebAssembly/binaryen#1153]
Repro: https://gist.github.com/kumpera/dbe1b3a1a28314160532dc07cf5b041a

Each file was individually compiled with:`emcc -g -Os  -s WASM=1 -s ALLOW_MEMORY_GROWTH=1  -s TOTAL_MEMORY=134217728 -c`

When linking all 3 object files binaryen asserts in:
`
Assertion failed: (isConcreteWasmType(block->type) && singleton->type == unreachable), function simplifyToContents, file XXX/emsdk/binaryen/master/src/ast/block-utils.h
`

I did some digging and the offending function _mono_thread_info_exit. The function body at the beginning of the Vacuum pass is:
```
(block
 (call $_mono_thread_info_detach)
 (block
  (block
   (drop
    (i32.const 0)
   )
   (drop
    (i32.const 0)
   )
   (block
    (set_local $0
     (get_global $STACKTOP)
    )
    (set_global $STACKTOP
     (i32.add
      (get_global $STACKTOP)
      (i32.const 16)
     )
    )
    (call $_g_log
     (i32.const 0)
     (i32.const 4)
     (i32.const 1524)
     (get_local $0)
    )
    (loop $while-in
     (br $while-in)
    )
   )
  )
 )
)
```

It seems somehow related to the infinite loop at the end since the problem goes away if I remove it from the code (line 75 of the gist).

**Comment by kripken** [comment]
Thanks, yeah, the issue is that infinite loop at the end (which has type `unreachable`), which was not properly handled by the inliner. By coincidence this bug was reported twice today, and there is already a fix for it in #1154 (the inlining fix there).

**Comment by kumpera** [comment]
Just updated my emsdk checkout with the latest binaryen and it's working!
Thanks a lot for such lightning fast response. 👍
